### ISSUE TITLE
Add the ability to trigger expansions using the CLI

### ISSUE THREAD

**Issue by phorcys420** [federico-terzi/espanso#1996]
### Description

I would like to be able to use triggers without the need for keyboard input.

---

Let's say we have the following matches :

```yml
matches:
  - trigger: hello
    replace: world
```

`espanso trigger hello` would output the result of the trigger to stdout, in this case it would output `world`.
`espanso trigger hello --type` would type out the result of the trigger.

### Motivation

- Ability to debug specific Espanso [extensions](https://espanso.org/docs/matches/extensions) without needing to use a GUI and regardless of the keyboard implementation during development. (would also allow for better unit testing in my opinion)

- Ability to integrate Espanso with other tools
For example, I might want to have both an Espanso trigger and a shortcut key, we could imagine a world where you define a hotkey in AutoHotKey (or on your Stream Deck) which calls espanso to write the trigger's output (that's where the `--type` flag comes in).

- Ability to reuse trigger output in a program/whatever.
Let's say you have a trigger that outputs a specific value related to what you do and you want to reuse that value without having to maintain it in multiple places, then you could defer to Espanso.

### Have you tried building it? Or would you like it?

- [ ] I'd be willing to contribute this feature to Espanso myself.

**Comment by smeech** [comment]
The CLI command is `espanso match exec -t <trigger>`.

I had similar thoughts (#1933). It's interesting that the syntax changed from Espanso v0.7.3 to c2.2.1, in that the `-t` was added, for no apparent reason, as without it the command does nothing. I'm not sure how `--arg <arg>` can be used.

The relevant code appears to be: https://github.com/espanso/espanso/blob/c219b56d45295200fed88546a20ed152fa14d5d0/espanso/src/cli/match_cli/exec.rs but it doesn't yield any clues that I can see.

**Comment by kud** [comment]
I think I've got exactly the same issue.

1. I'm confused about the need for `-t`.
2. Executing the trigger not only gives the result of the trigger but also executes it by adding the result to your cursor.

Fun fact: I was trying to execute this via a script in Raycast, and I ended up with a loop because by executing this command, it also triggered Espanso itself 🥲

@federico-terzi any thought on that? I definitely need to get only the result of the trigger without executing it on the cursor, for my Raycast extension.

**Comment by phorcys420** [comment]
Hey all, I forgot to answer, I think this is nice but not really what I want, I'd like to be able to get the output of the trigger without Espanso typing it.

Maybe that's what `-t` was intended for in the first place? (as in "enable typing")

**Comment by smeech** [comment]
I think we're on the same page, but I _don't_ see Espanso typing the trigger when called with `espanso match exec -t :espanso` attached to a system keyboard combination, but it *does* try to _delete_ the trigger, which results in it removing the previous text.

Are you seeing something different?

**Comment by kud** [comment]
https://github.com/espanso/espanso/assets/655838/66ba4b12-d70e-4aae-a784-b1f992c9ce3e

As you can see in this video, executing the command returns the output AND also replace the current cursor.

**Comment by kud** [comment]
All I want in my case is to execute the trigger AND just return the output, but not actually really triggers the espanso match.

Why? Because I'm the owner of the extension espanso for Raycast and what I want to do is to execute the trigger and get the result WITHOUT writting in the cursor the result, because it actually creates a loop 😭

In other words, I would like a command which processes the match and gives me the result, but does not execute the trigger.

**Comment by smeech** [comment]
Do you get the same in other environments, e.g. terminal? Is this a Raycast issue, rather than Espanso?
Which version of Raycast? I see there have been some clipboard improvements in the last few months.
What exactly is your trigger snippet?

**Comment by kud** [comment]
It's definitely not a raycast issue in any case. I'm using `zx` to run the command, but even without zx, as you can see, it's triggered in iTerm.app.

Same in Terminal.app.

![screenshot-2024-07-11-16-14-50@2x](https://github.com/espanso/espanso/assets/655838/d0f74945-0e37-4ad3-a817-58824928eab1)


I'm pretty sure what's happening is the command prints the result and also execute the trigger which then replaces the current content of the cursor by the new one.

```
  - trigger: "??nohello"
    label: "No Hello"
    replace: "https://nohello.net/"
```

**Comment by smeech** [comment]
Very odd!

**Comment by kud** [comment]
I'm not so surprised to be honest.

The command triggers the match so it returns the process of the match + executes it on the cursor itself.

**Comment by federico-terzi** [comment]
It's been a long time since I worked on this area, so my memory is quite fuzzy. I do remember that in Espanso v2, there's the concept of "compensation" which is used to undo the trigger and to move the cursor to the appropriate location: https://github.com/espanso/espanso/blob/dev/espanso-engine/src/event/effect.rs#L23C12-L31

By dispatching a `TextInjectRequest` _without_ `TriggerCompensation`, you should be able to inject the trigger without causing the cursor to move or previous text to be deleted.

I would also try to look into how Espanso handles the Search Bar injections, as in that case there's no compensation in some cases (eg. when triggering it with Alt+Space)

Hope this helps!

**Comment by kud** [comment]
Hey @federico-terzi, thank you so much for your time and your explanation.

It means that we probably need either to fix the `match exec` command or creating another command then. It depends what you guys expect from `match exec`.

All I know is I would like to improve the cli to be able to use it via a script like shell, zx, etc.

**Comment by AucaCoyan** [comment]
> It depends what you guys expect from match exec.

`espanso match exec` is a bit odd, it doesn't [have a helpful help page](https://espanso.org/docs/command%20lIne/cli_list/#exec), and that is a sign of a API that could be improved.
If we prevent the expansion by looking up `espanso match exec` before any trigger, you _could_ type `espanso match exec <trigger>` and it wouldn't expand. This has the small problem that it **has** to be typed, `espanso match exec` can't be aliased otherwise the lookup wouldn't work. 
It's a bit of a mouthful typing 3 words before a trigger, which in some cases is a bit long, so I propose moving this API into a different command:

```console
$ espanso try
Tries a match

Usage: espanso try <TRIGGER>

Arguments:
  <TRIGGER>  

Options:
  -h, --help  Print help
```

I don't know if:
```
        --arg <arg>...         Specify also an argument for the expansion, following the 
                               --arg 'name=value' format. 
                               You can specify multiple args.
```
is even necessary, or needed in the first deliver of this cmd. We could deliver it in a second phase.

**Comment by smeech** [comment]
Happy to shorten the command, although I suspect most invocations will be from a script or system keybinding command, so it doesn't matter too much. It does fit the syntax of [`espanso match`](https://espanso.org/docs/command%20lIne/cli_list/#match), although `--arg <arg>` and `-t` could be dispensed with.

I *would* like https://github.com/espanso/espanso/issues/1933 addressed in this context, which might make `-t` useful!

**Comment by kud** [comment]
As long as I can use the result in a script, I'm happy to have any command. Thanks for looking at it @smeech !

**Comment by AucaCoyan** [comment]
'kay, so it would be:

```console
espanso match exec <trigger>
```

and the features:
- espanso will look up the trigger with the key `<trigger>`.
- espanso should not trigger the `<trigger>` part (it defeats the purpose of triggering via the terminal)
- also, it doesn't delete the number of chars of the trigger, causing the issue #1555 
- we are not going to provide other arguments, `--args` or `-t`, at least for now.

I think that covers a bunch of cases, but there is a scenario I still have some doubts, how do you trigger something with the Alt key, or F1, how do you type that into the terminal?

For reference, `xdotool` uses a notation specifying keys as text [man page](https://www.mankier.com/1/xdotool#Keyboard_Commands)

**Comment by smeech** [comment]
> * espanso should not trigger the `<trigger>` part

I think you mean Espanso should not print the `trigger:` value, _or_ attempt to delete it.

> how do you trigger something with the Alt key, or F1

By assigning `espanso match exec <trigger>` to a _system_ keybinding, as I outline in https://github.com/espanso/espanso/issues/1933. E.g. (my current test): 

![Image](https://github.com/user-attachments/assets/f8c44739-952f-4c0d-a1e4-e9d9293affda)

It's necessary to test this in an editor _with some preceding text_ to see the current problem.

**Comment by AucaCoyan** [comment]
Ah! I thought it was a required to solve this issue that you can send keys via the terminal (kinda like xdotool).
If that is not a blocker, it's a small fix.
I have to solve the "do not delete pls" and the rest is deleting code

**Comment by smeech** [comment]
> I thought it was a required to solve this issue that you can send keys via the terminal (kinda like xdotool).

I don't think so. My current _workaround_ (because `espanso match exec` doesn't work usefully) is to generate a trigger value with `xdotool` so that Espanso has something to delete. If we can rewrite `espanso match exec` to work without generating or deleting the `trigger:` value, that's not necessary, and enables the code to work on Windows too (which doesn't have `xdotool`, but does have system keybindings).

By the way, I am pleased to find that assigning an `espanso match exec` command to a system keybinding as shown in https://github.com/espanso/espanso/issues/1996#issuecomment-2599096269, _does_ follow an app-specific configuration include/exclude rule!
